Update astc to 4.8.0, enable SSE41 decode, and fix build issues for physical iOS targets

SRSaunders opened the pull request:
## Description

The new astc library from commit https://github.com/KhronosGroup/Vulkan-Samples/commit/7017dc07869aa0fd0eba1fd440010f10d56407b2 does not generate a static (.a) library when cross-compiling from x86_64 hosts to arm64 targets. This PR fixes this build problem by making sure "neon" is set for all iOS physical (not simulator) targets, and "native" otherwise.

This PR also removes a previous workaround for macOS/iOS arm64 targets since the original defect is now fixed.  See fixed issue https://github.com/ARM-software/astc-encoder/issues/458.

Fixes #1294.

Tested on macOS x86_64 (Intel iMac) and arm64 (M1 MacBook Air) hosts building for native macOS execution.  Also tested on macOS x86_64 hosts (Intel iMac) building for iOS physical (arm64 iPhone 15) and simulator (x86_64 host) targets.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

 If this PR contains framework changes:
 - [ ] I did a full batch run using the `batch` command line argument to make sure all samples still work properly

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [ ] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the readme of the folder that the sample belongs to [e.g. api samples readme](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/samples/api/README.adoc)
- [ ] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/conditional_rendering)
- [ ] For new samples, I have added a link to the [Antora navigation](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/antora/modules/ROOT/nav.adoc) so that the sample will be listed at the Vulkan documentation site

SRSaunders committed:
Always build astc for arm64 when target is iOS physical device

SRSaunders committed:
Remove previous astc arm64 workaround since original defect is now fixed

gpx1000 commented on third_party/CMakeLists.txt:
```diff
@@ -308,8 +308,13 @@ if (NOT (DEFINED ASTC_ARCH))
+    # Always build for ARM when compiling or cross-compiling to iOS physical devices, otherwise fall back to native as above
     if (APPLE)
-        set(ASTC_ARCH NATIVE)
+		if(IOS AND ${CMAKE_OSX_SYSROOT} STREQUAL "iphoneos")
```
Does this work for iPad?

Great catch by the way!

SRSaunders replied:
@gpx1000 yes it should work for iPad as is (which is an iOS target from a developer perspective).  As far as I know, the valid Apple target platforms (cmake SysRoots) are:

![Screenshot 2025-03-11 at 5 46 38 PM](https://github.com/user-attachments/assets/92442305-7532-445f-95bf-84c561daf72c)

This project supports three of them: _macosx_, _iphoneos_, and _iphonesimulator_.  While I have not tested on a physical iPad (don't have one), I have tested using a virtual iPad using the _iphonesimulator_ target, and that works fine.  If you have a physical iPad, perhaps you could test and let me know if there are any issues.  Thx.

Note that _iphoneos_ is already used in **.../Vulkan-Samples/app/CMakeLists.txt**, and _iphonesimulator_ is used in **.../Vulkan-Samples/bldsys/cmake/global_options.cmake** to test for iOS physical devices vs. sim  devices.

gpx1000 replied:
Yep, that looks right, the CMake docs say that it could be appleTV, iOS, MacOS, WatchOS, or visionOS etc.  I'm okay with the changes here thanks!

SRSaunders replied:
Thanks for reviewing!

SRSaunders committed:
Update astc submodule to version 4.8.0

SRSaunders committed:
Add SSE41 decoding option for all x86 platforms (Win, Linux, macOS)

SRSaunders commented:
As per discussion with @gpx1000 and @gary-sweet I have updated the astc library to version 4.8.0 to pick up macOS fix https://github.com/ARM-software/astc-encoder/commit/71a55e7697a72df9e004a8212288f1ef71572f9e.

In addition, I have added an enhancement to enable SSE41 decoding for all x86 platforms (Win, Linux, macOS) as a fallback option if AVX2 is not available.  This was likely not enabled previously due to a mismatch in naming convention for the target library which should be **libastcdec-sse4.1-static.a** and not **libastcdec-sse41-static.a**.  Subtle difference with "." in name.

SRSaunders commented:
@SaschaWillems, if you review this PR please note that I could possibly collapse the APPLE vs. Other logic for **astc** vector instruction selection into a single cmake code block.  However, I was reluctant to do this without more info since I don't really understand how Android is built.  I thought it would be a normal C++ cross-compilation target from x86 to arm64, but looking at the CI build log for Android it seems more complicated than that.

In fact I see that many libraries are not built during CI, with the message: `not building target library <...> because static libraries are not build by default`

Also the sample targets are not built and show the message: `not building target <...> because no targets are specified and library build output file is null`

Is this expected behaviour for the Android CI build?

In any case, getting back to **astc**, if Android is a normal cross-compilation target, I could simply force NEON on for Android builds just like iOS physical target builds, and then collapse the APPLE and non-APPLE x86 cmake code blocks into one.  However, if Android uses a self-hosted or virtual self-hosted dev environment then the answer may be different. Would you please let me know how Android is built: cross compilation or virtual self-hosted?  Thanks.

Note this question is only about possible code minimization, and this PR will work as-is without modification.

SaschaWillems commented:
I know close to noting about our build setup on Android, so sadly can't answer your question.

All I can say is that it builds for me locally and that samples that use ASTC do work as expected. So from that pov, this PR is fine.

Sadly samples on Android are so broken, that it's nigh impossible to test them right, see #1299, though that's not an issue fo this PR.

We plan on removing ASTC at some point though, as there are better options that won't take ages to transcode textures to native formats on desktop.

I guess this makes #1300 obsolete then?

SRSaunders commented:
Thanks @SaschaWillems for your answer.  I will leave things alone since it is working for you on Android, etc.

> I guess this makes https://github.com/KhronosGroup/Vulkan-Samples/pull/1300 obsolete then?

Yes, I suppose it does.  However, I would like @gpx1000 and @gary-sweet to review this PR before making that decision.  They indicated they may want to return to a stable base (astc 4.7) before any further changes.  However, if they agree with taking a single step to astc 4.8, then I would be fine with that too.

SRSaunders committed:
Merge branch 'main' into astc-ios-fix

SRSaunders committed:
Re-update astc submodule to version 4.8.0 following rebase

SRSaunders committed:
Restore ray_tracing_reflection's README changes following rebase

SRSaunders commented:
Had to rebase following merge of #1300 due to astc downgrade merge conflict.

gary-sweet reviewed (approved):
Doesn't seem to break anything for me, though I can't comment on the Mac stuff.

marty-johnson59 commented:
Merging with Steve's approval (per discussion on 24 March Samples call)
